strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,363 @@
+
+
+# TODO: update embedding class in type hint
+class VGBS:
+    r"""Create a variational GBS model for optimization and machine learning.
+
+    An input adjacency matrix :math:`A` can be varied using:
+
+    .. math::
+
+        A(\theta) = W(\theta) A W(\theta)
+
+    with :math:`W` a diagonal matrix of weights that depend on a set of parameters :math:`\theta`.
+
+    By varying :math:`\theta`, the distribution of samples from GBS can be preferentially biased
+    towards solving stochastic optimization and unsupervised machine learning problems.
+
+    The above variational model can be used in both the threshold and PNR modes of GBS, which
+    is specified using the ``threshold`` flag. An initial value for the mean number of clicks
+    (if ``threshold=True``) or mean number of photons (if ``threshold=False``) is also required.
+    The initial value ``n_mean`` should be set high since varying :math:`\theta` can only lower
+    the mean number of clicks or photons.
+
+    The mapping from :math:`\theta` to :math:`W(\theta)` is specified using the ``embedding``
+    argument.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(4, 0.7, seed=1967)
+    >>> a = nx.to_numpy_array(g)
+    >>> embedding = Exp()  # TODO does this need updating?
+    >>> vgbs = VGBS(a, 3, embedding, True)
+    >>> params = np.array([0.05, 0.1, 0.02, 0.01])
+    >>> vgbs.A(params)
+    array([[0.        , 0.86070798, 0.93239382, 0.94176453],
+           [0.86070798, 0.        , 0.88692044, 0.89583414],
+           [0.93239382, 0.88692044, 0.        , 0.97044553],
+           [0.94176453, 0.89583414, 0.97044553, 0.        ]])
+    >>> vgbs.n_mean(params)
+    1.8906000271819803
+    >>> vgbs.generate_samples(params, 2)
+    array([[0, 0, 1, 1],
+           [1, 1, 0, 0]])
+
+    Args:
+        A (array): the input adjacency matrix :math:`A`
+        n_mean (float): the initial mean number of clicks or photons
+        embedding: the method of converting from trainable parameters :math:`\theta` to
+            :math:`W(\theta)` #TODO: any update from other PR?
+        threshold (bool): determines whether to use GBS in threshold or PNR mode
+        samples (array): an optional array of samples from :math:`A` used to speed up gradient
+            calculations #TODO: more info
+    """
+    def __init__(
+        self,
+        A: np.ndarray,
+        n_mean: float,
+        embedding,
+        threshold: bool,
+        samples: Optional[np.ndarray] = None,
+    ):
+        if not np.allclose(A, A.T):
+            raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+        self.A_init = A
+        self.embedding = embedding
+        self.threshold = threshold
+        self.A_init_scaled = rescale_adjacency(A, n_mean, threshold)
+        self.n_modes = len(A)
+        self._A_init_samples = None
+        if samples:
+            self._add_A_init_samples(samples)
+
+    def W(self, params: np.ndarray) -> np.ndarray:
```

Comment:
Here we have to make a decision regarding convention: the diagonal entries of `W` can be either `w` or `sqrt(w)`. In the paper, we use `sqrt(w)` and so in the gradient formulas only the `w` comes out. If we use `w` instead, gradient formulas will appear with `w^2`

Reply:
Oh nice, I have changed to use square root - which also makes this method more nontrivial :laughing: